tests/KeyOfArrayTest.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psalm\Tests;
+
+use Psalm\Tests\Traits\InvalidCodeAnalysisTestTrait;
+use Psalm\Tests\Traits\ValidCodeAnalysisTestTrait;
+
+class KeyOfArrayTest extends TestCase
+{
+    use InvalidCodeAnalysisTestTrait;
+    use ValidCodeAnalysisTestTrait;
+
+    /**
+     * @return iterable<string,array{code:string,assertions?:array<string,string>,ignored_issues?:list<string>,php_version?:string}>
+     */
+    public function providerValidCodeParse(): iterable
+    {
+        return [
+            'keyOfListClassConstant' => [
+                'code' => '<?php
+                    class A {
+                        const FOO = [
+                            \'bar\'
+                        ];
+                        /** @return key-of<A::FOO> */
+                        public function getKey() {
+                            return 0;
+                        }
+                    }
+                '
+            ],
+            'keyOfAssociativeArrayClassConstant' => [
+                'code' => '<?php
+                    class A {
+                        const FOO = [
+                            \'bar\' => 42
+                        ];
+                        /** @return key-of<A::FOO> */
+                        public function getKey() {
+                            return \'bar\';
+                        }
+                    }
+                '
+            ],
+            'allKeysOfAssociativeArrayPossible' => [
+                'code' => '<?php
+                    class A {
+                        const FOO = [
+                            \'bar\' => 42,
+                            \'adams\' => 43,
+                        ];
+                        /** @return key-of<A::FOO> */
+                        public function getKey(bool $adams) {
+                            if ($adams) {
+                                return \'adams\';
+                            }
+                            return \'bar\';
+                        }
+                    }
+                '
+            ],
+            'keyOfAsArray' => [
+                'code' => '<?php
+                    class A {
+                        /** @var array */
+                        const FOO = [
+                            \'bar\' => 42,
+                            \'adams\' => 43,
+                        ];
+                        /** @return key-of<self::FOO>[] */
+                        public function getKey(bool $adams) {
+                            return array_keys(self::FOO);
+                        }
+                    }
+                '
+            ],
+            'keyOfArrayLiteral' => [
+                'code' => '<?php
+                    /**
+                     * @return key-of<array<int, string>>
```

Comment:
Is this something we want to accept? Does it have any concrete use as opposed to returning `int`?

Reply:
I added it, to test if `key-of<properties-of<Foo>>` will work, if I rewrote the related PR to return a keyed-array  instead of only the names.

Additionally this works in TypeScript as well.

Reply:
Fine by me :)